README.md
[![Build Status](https://github.com/Polymer/prpl-server/actions/workflows/tests.yml/badge.svg)](https://github.com/Polymer/prpl-server/actions/workflows/tests.yaml?query=branch%3Amain)
[![Published on npm](https://img.shields.io/npm/v/prpl-server.svg?logo=npm)](https://www.npmjs.com/package/prpl-server)

# prpl-server

An HTTP server for Node designed to serve [PRPL](https://developers.google.com/web/fundamentals/performance/prpl-pattern/) apps in production.

## Deprecation
> ⚠️IMPORTANT⚠️ `prpl-server` is deprecated and is no longer recommended. Reports and PRs for critical bugs and security issues will be accepted, but we will no longer accept new feature requests or PRs.
> 
> For differential serving, we now recommend a simple two-build configuration using `nomodule` for client-side capability sniffing. See https://jasonformat.com/modern-script-loading/ for more details on this pattern.
>
> If you are looking for a modern *development* sever, we recommend [@web/dev-server](https://modern-web.dev/docs/dev-server/overview/).
> 
> In addition, note that Chromium is [considering](https://groups.google.com/a/chromium.org/g/blink-dev/c/K3rYLvmQUBY/m/vOWBKZGoAQAJ?pli=1) removing support for Server Push in a future version. Consider using [`<link rel="preload">`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) tags as a simpler alternative with similar performance benefits.

## Contents
- [Usage](#usage)
  - [As a binary](#as-a-binary)
  - [As a library](#as-a-library)
- [Differential Serving](#differential-serving)
  - [Builds](#builds)
  - [Capabilities](#capabilities)
- [Entrypoint](#entrypoint)
- [Base paths](#base-paths)
- [HTTP/2 Server Push](#http-2-server-push)
  - [Push manifest](#push-manifest)
  - [Link preload headers](#link-preload-headers)
  - [Testing push locally](#testing-push-locally)
- [Service Workers](#service-workers)
  - [Scope header](#scope-header)
  - [404 handling](#404-handling)
- [HTTPS](#https)
- [Caching](#caching)
- [HTTP Errors](#http-errors)
- [Rendering for Bots](#rendering-for-bots)
- [Google App Engine Quickstart](#google-app-engine-quickstart)

## Usage

### As a binary
```sh
$ npm install -g prpl-server
$ prpl-server --root . --config polymer.json
```

### As a library

```sh
$ npm install --save prpl-server
```

```js
prpl = require('prpl-server');
express = require('express');

const app = express();

app.get('/api/launch', (req, res, next) => res.send('boom'));

app.get('/*', prpl.makeHandler('.', {
  builds: [
    {name: 'modern', browserCapabilities: ['es2015', 'push']},
    {name: 'fallback'},
  ],
}));

app.listen(8080);
```

## Differential Serving

Modern browsers offer great features that improve performance, but most applications need to support older browsers too. prpl-server can serve different versions of your application to different browsers by detecting browser capabilities using the user-agent header.

### Builds

prpl-server understands the notion of a *build*, a variant of your application optimized for a particular set of browser capabilities.

Builds are specified in a JSON configuration file. This format is compatible with [`polymer.json`](https://www.polymer-project.org/2.0/docs/tools/polymer-json), so if you are already using polymer-cli for your build pipeline, you can annotate your existing builds with browser capabilities, and copy the configuration to your server root. prpl-server will look for a file called `polymer.json` in the server root, or you can specify it directly with the `--config` flag.


In this example we define two builds, one for modern browsers that support ES2015 and HTTP/2 Push, and a fallback build for other browsers:

```
{
  "entrypoint": "index.html",
  "builds": [
    {"name": "modern", "browserCapabilities": ["es2015", "push"]},
    {"name": "fallback"}
  ]
}
```

### Capabilities

The `browserCapabilities` field defines the browser features required for that build. prpl-server analyzes the request user-agent header and picks the best build for which all capabilities are met. If multiple builds are compatible, the one with more capabilities is preferred. If there is a tie, the build that comes earlier in the configuration file wins.

You should always include a fallback build with no capability requirements. If you don't, prpl-server will warn at startup, and will return a 500 error on entrypoint requests to browsers for which no build can be served.

The following keywords are supported. See also the [browser-capabilities](https://github.com/Polymer/tools/tree/master/packages/browser-capabilities) library which prpl-server uses.

| Keyword       | Description
| :----         | :----
| push          | [HTTP/2 Server Push](https://developers.google.com/web/fundamentals/performance/http2/#server-push)
| serviceworker | [Service Worker API](https://developers.google.com/web/fundamentals/getting-started/primers/service-workers)
| modules       | [JavaScript Modules](https://www.chromestatus.com/feature/5365692190687232) (including dynamic `import()` and `import.meta`)
| es2015        | [ECMAScript 2015 (aka ES6)](https://developers.google.com/web/shows/ttt/series-2/es2015)
| es2016        | ECMAScript 2016
| es2017        | ECMAScript 2017
| es2018        | ECMAScript 2018

## Entrypoint

In the [PRPL pattern](https://developers.google.com/web/fundamentals/performance/prpl-pattern/), the *entrypoint* is a small HTML file that acts as the application bootstrap.

prpl-server will serve the entrypoint from the best compatible build from `/`, and from any path that does not have a file extension and is not an existing file.

prpl-server expects that each build subdirectory contains its own entrypoint file. By default it is `index.html`, or you can specify another name with the `entrypoint` configuration file setting.

Note that because the entrypoint is served from many URLs, and varies by user-agent, cache hits for the entrypoint will be minimal, so it should be kept as small as possible.

## Base paths

Since prpl-server serves resources from build subdirectories, your application source can't know the absolute URLs of build-specific resources upfront.

For most documents in your application, the solution is to use relative URLs to refer to other resources in the build, and absolute URLs to refer to resources outside of the build (e.g. static assets, APIs). However, since the *entrypoint* is served from URLs that do not match its location in the build tree, relative URLs will not resolve correctly.

The solution we recommend is to place a [`<base>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base) tag in your entrypoint to anchor its relative URLs to the correct build subdirectory, regardless of the URL the entrypoint was served from. You may then use relative URLs to refer to build-specific resources from your entrypoint, as though you were in your build subdirectory. Put `<base href="/">` in your source entrypoint, so that URLs resolve when serving your source directly during development. In your build pipeline, update each entrypoint's base tag to match its build subdirectory (e.g. `<base href="/modern/">`).

If you are using polymer-cli, set `{"autoBasePath": true}` in your `polymer.json` to perform this base tag update automatically.

Note that `<base>` tags only affect relative URLs, so to refer to resources outside of the build from your entrypoint, use absolute URLs as you normally would.

## HTTP/2 Server Push

Server Push allows an HTTP/2 server to preemptively send additional resources alongside a response. This can improve latency by eliminating subsequent round-trips for dependencies such as scripts, CSS, and HTML imports.


### Push manifest

prpl-server looks for a file called `push-manifest.json` in each build subdirectory, and uses it to map incoming request paths to the additional resources that should be pushed with it. The original push manifest file format is described [here](https://github.com/GoogleChrome/http2-push-manifest). Tools for generating a push manifest include [http2-push-manifest](https://github.com/GoogleChrome/http2-push-manifest) and [polymer-cli](https://github.com/Polymer/polymer-cli).

Each key in the push manifest is a regular expression pattern that will be matched against the incoming request path. Patterns are forced to match exactly (e.g. `foo.html` is equivalent to `^foo.html$`). You can use wildcard patterns to push resources for client-side application routes (e.g. `/articles/.*`). In the case of the entrypoint, the resolved filename (e.g. `index.html`) is used as a key to the push manifest, in addition to the application route.

Resources in the push manifest can be specified as absolute or relative paths. Absolute paths are interpreted relative to the server root directory. Relative paths are interpreted relative to the location of the push manifest file itself (i.e. the build subdirectory), so that they do not need to know which build subdirectory they are being served from. Push manifests generated by `polymer-cli` always use relative paths.

### Link preload headers

prpl-server is designed to be used behind an HTTP/2 reverse proxy, and currently does not generate push responses itself. Instead it sets [preload link](https://w3c.github.io/preload/#server-push-http-2) headers, which are intercepted by cooperating reverse proxy servers and upgraded into push responses. Servers that implement this upgrading behavior include [Apache](https://httpd.apache.org/docs/trunk/mod/mod_http2.html#h2push), [nghttpx](https://github.com/nghttp2/nghttp2#nghttpx---proxy), and [Google App Engine](https://cloud.google.com/appengine/).

If a build with a push manifest is served to a browser that does not support push according to the [browser-capabilities](https://github.com/Polymer/tools/tree/master/packages/browser-capabilities) support matrix, then a `nopush` attribute is added to the generated preload link headers.

### Testing push locally

To confirm your push manifest is working during local development, you can look for `Link: <URL>; rel=preload` response headers in your browser dev tools.

To see genuine push locally, you will need to run a local HTTP/2 reverse proxy such as [nghttpx](https://github.com/nghttp2/nghttp2#nghttpx---proxy):

- Install nghttpx ([Homebrew](http://brewformulas.org/Nghttp2), [Ubuntu](http://packages.ubuntu.com/zesty/nghttp2), [source](https://github.com/nghttp2/nghttp2#building-from-git)).
- Generate a self-signed TLS certificate, e.g. `openssl req -newkey rsa:2048 -x509 -nodes -keyout server.key -out server.crt`
- Start prpl-server (assuming default `127.0.0.1:8080`).
- Start nghttpx: `nghttpx -f127.0.0.1,8443 -b127.0.0.1,8080 server.key server.crt --no-ocsp`
- Visit `https://localhost:8443`. In Chrome, Push responses will show up in the Network tab as Initiator: Push / Other.

Note that Chrome will not allow a service worker to be registered over HTTPS with a self-signed certificate. You can enable [chrome://flags/#allow-insecure-localhost](chrome://flags/#allow-insecure-localhost) to bypass this check. See [this page](https://www.chromium.org/blink/serviceworker/service-worker-faq) for more tips on developing service workers in Chrome.

## Service Workers

### Scope header
prpl-server sets the [`Service-Worker-Allowed`](https://www.w3.org/TR/service-workers-1/#service-worker-allowed) header to `/` for any request path ending with `service-worker.js`. This allows a service worker served from a build subdirectory to be registered with a scope outside of that directory, e.g. `register('service-worker.js', {scope: '/'})`.

### 404 handling

prpl-server automatically serves a tiny self-unregistering service worker for any request path ending with `service-worker.js` that would otherwise have had a `404 Not Found` response. To disable this behavior, set `unregisterMissingServiceWorkers: false` in your configuration file.

This can be useful when the location of a service worker has changed, as it will prevent clients from getting stuck with an old service worker indefinitely.

This problem arises because when a service worker updates, a `404` is treated as a failed update. It does not cause the service worker to be unregistered. See [w3c/ServiceWorker#204](https://github.com/w3c/ServiceWorker/issues/204) for more discussion of this problem.

## HTTPS

Your apps should always be served over HTTPS. It protects your user's data, and is *required* for features like service workers and HTTP/2.

If the `--https-redirect` flag is set, prpl-server will redirect all HTTP requests to HTTPS. It sends a `301 Moved Permanently` redirect to an `https://` address with the same hostname on the default HTTPS port (443).

prpl-server trusts [`X-Forwarded-Proto`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto) and [`X-Forwarded-Host`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host) headers from your reverse proxy to determine the client's true protocol and hostname. Most reverse proxies automatically set these headers, but if you encounter issues with redirect loops, missing or incorrect `X-Forwarded-*` headers may be the cause.

You should always use `--https-redirect` in production, unless your reverse proxy already performs HTTPS redirection.

## Caching

By default, prpl-server sets the [`Cache-Control`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) header to `max-age=60` (1 minute), except for the entrypoint and service worker which gets `max-age=0`. [`ETag`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) headers are also sent, so resources that have not changed on the server can be re-validated efficiently.

To change this default for non-entrypoint resources, set the `cacheControl` property in your configuration file, or the `--cache-control` command-line flag, to the desired `Cache-Control` header value. You may want to set `--cache-control=no-cache` during development.

For more advanced caching behavior, [use prpl-server as a library](#as-a-library) with Express and register a middleware that sets the `Cache-Control` header before registering the prpl-server middleware. If prpl-server sees that the `Cache-Control` header has already been set, it will not modify it. For example, to set year-long caching for images:

```js
app.get('/images/*', (req, res, next) => {
  res.setHeader('Cache-Control', 'public, max-age=31536000');
  next();
});

app.get('/*', prpl.makeHandler('.', config))
```

Choosing the right cache headers for your application can be complex. See [*Caching best practices & max-age gotchas*](https://jakearchibald.com/2016/caching-best-practices/) for one starting point.

## HTTP Errors

By default, if a `404 Not Found` or other HTTP server error occurs, prpl-server will serve a minimal `text/plain` response. To serve custom errors, [use prpl-server as a library](#as-a-library) with Express, set `forwardErrors: true` in your configuration object, and register an [error-handling middleware](http://expressjs.com/en/guide/error-handling.html) after registering the prpl-server handler:

```js
app.get('/*', prpl.makeHandler('.', {
  builds: [ ... ],
  forwardErrors: true
}));

app.use((err, req, res, next) => {
  if (err.status === 404) {
    res.status(404).sendFile('my-custom-404.html', {root: rootDir});
  } else {
    next();
  }
});
```

## Rendering for Bots

Many bots don't execute JavaScript when processing your application. This can cause your application to not render correctly when crawled by some search engines, social networks, and link rendering bots.

One solution to this problem is [Rendertron](https://github.com/GoogleChrome/rendertron). Rendertron is a server which runs headless Chrome to render and serialize web pages for these bots, so all the content is contained in one network request. Use the `--bot-proxy` flag to instruct prpl-server to proxy requests from a known list of bots through a Rendertron server.

Note that you can also use the [Rendertron middleware](https://github.com/GoogleChrome/rendertron/tree/master/middleware) directly if you have a custom Express server.

## Google App Engine Quickstart

[Google App Engine](https://cloud.google.com/appengine/) is a managed server platform that [supports Node.js](https://cloud.google.com/nodejs/). You can deploy prpl-server to App Engine [standard environment](https://cloud.google.com/appengine/docs/standard/nodejs/) with a few steps:

1. Follow [these instructions](https://cloud.google.com/appengine/docs/standard/nodejs/quickstart) to set up a Google Cloud project and install the Google Cloud SDK. As instructed, run the `gcloud init` command to authenticate and choose your project ID.

2. `cd` to the directory you want to serve (e.g. your app's `build/` directory if you are using polymer-cli).

3. Run `npm init` and follow the prompts to create your `package.json`.

4. Run `npm install --save prpl-server` to add prpl-server as a dependency.

5. Edit your `package.json` to add a `start` script. This is the command App Engine runs when your app starts. Configure `prpl-server` to listen on all hosts, and to redirect HTTP connections to HTTPS. You should also specify the version of Node your app requires via the `engines` section.

```json
{
  "scripts": {
    "start": "prpl-server --host 0.0.0.0 --https-redirect"
  },
  "engines": {
    "node": "8.x.x"
  }
}
```

6. Create an `app.yaml` file. This tells App Engine that you want to use Node.js in the App Engine standard environment:

```yaml
runtime: nodejs8
```

7. Run `gcloud app deploy` to deploy to your App Engine project. `gcloud` will tell you the URL your app is being served from. For next steps, check out the Node.js in the App Engine standard environment [documentation](https://cloud.google.com/appengine/docs/standard/nodejs/).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

<!-- ## Unreleased -->
<!-- Add new, unreleased changes here. -->

## [1.4.3] 2021-09-22
- `prpl-server` is now deprecated. A warning will be printed when installing
  from NPM, as well as each time the server is launched. Please see
  https://github.com/Polymer/prpl-server#deprecation for details and recommended
  alternatives.
- Updated dependencies.
- Updated some repo links (`prpl-server-node` -> `prpl-server`).

## [1.4.2] 2020-08-12
- Update dependencies.
- Clean up unnecessary NPM package files.

## [1.4.1] 2020-08-07
- Clean up unnecessary dependencies.

## [1.4.0] 2019-01-07
- Don't cache response for SW.

## [1.3.0] 2018-12-06
- Support crossorigin setting in push-manifest.

## [1.2.0] 2018-05-22
- Add `as=fetch` as a valid value for preload headers.

## [1.1.0] 2018-04-23
- Update browser-capabilities to pick up the latest user agent information for modules and service workers. Support for dynamic `import()` and `import.meta` are now requirements for the `modules` capability.
- If a build with a push manifest is served to a browser that does not support push (according to browser-capabilities), then we will still set preload headers, but with the `nopush` attribute set.

## [1.0.0] 2017-10-31
- Add `forwardErrors` option to pass 404s and other HTTP errors down to the next Express error-handling middleware.
- Recommend `npm` instead of `yarn` and switch to `npm` lock file.
- Check file existence asynchronously so the event loop is not blocked.

## [0.11.0] 2017-10-23
- Add `unregisterMissingServiceWorkers` option (default true) which serves a tiny self-unregistering service worker for would-be 404 service worker requests, to prevent clients from getting stuck with invalid service workers indefinitely.

## [0.10.2] 2017-10-18
- Require latest browser capabilities, which removes Firefox from push capable browsers due to https://bugzilla.mozilla.org/show_bug.cgi?id=1409570.
- Bump Yarn lock dependencies.

## [0.10.1] 2017-09-12
- Check the original URL path against the push manifest in addition to the resolved filename. This allows mapping application route patterns to push resources.

## [0.10.0] 2017-09-11
- Push manifest keys are now regular expression patterns instead of exact paths.
- The `Cache-Control` header is now set to 1 minute by default (except for the entrypoint). Added the `cacheControl` config property and `--cache-control` flag to override.

## [0.9.0] 2017-08-23
- Add `--bot-proxy` flag to proxy requests from bots through [Rendertron](https://github.com/GoogleChrome/rendertron).

## [0.8.0] 2017-08-09
- Switch to https://github.com/Polymer/browser-capabilities library.
- Add `modules` capability.
- Declare TypeScript typings in package.

## [0.7.0] 2017-05-30
- Add `--version` flag.

## [0.6.0] 2017-05-23
- Relative push manifest paths are now interpreted as relative to the location of the push manifest file itself. Previously they were always interpreted as relative to the server root.
- Extra safeguard against directory traversal attacks.

## [0.5.0] 2017-05-22
- Add `serviceworker` to browser capability detection.

## [0.4.0] 2017-05-19
- Add HTTP to HTTPS redirection.

## [0.3.0] 2017-05-18
- The commandline server now compresses responses.
- Fixed Windows bugs; now tested on AppVeyor.

## [0.2.0] 2017-05-16
- Initial release.


src/test/server_test.ts
/**
 * @license
 * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from 'chai';
import {server} from '../server';

suite('prpl server', function() {
  test('is created', () => {
    const app = server();
    assert.isOk(app);
  });
});


src/prpl.ts
/**
 * @license
 * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import * as capabilities from 'browser-capabilities';
import * as express from 'express';
import * as fs from 'fs';
import * as http from 'http';
import * as httpErrors from 'http-errors';
import * as path from 'path';
import * as send from 'send';
import * as statuses from 'statuses';
import * as url from 'url';

import * as push from './push';

export interface Config {
  // The Cache-Control header to send for all requests except the entrypoint.
  //
  // Defaults to `max-age=60`.
  cacheControl?: string;

  // If `true`, when a 404 or other HTTP error occurs, the Express `next`
  // function will be called with the error, so that it can be handled by
  // downstream error handling middleware.
  //
  // If `false` (or if there was no `next` function because Express is not
  // being used), a minimal `text/plain` error will be returned.
  //
  // Defaults to `false`.
  forwardErrors?: boolean;

  // Serves a tiny self-unregistering service worker for any request path
  // ending with `service-worker.js` that would otherwise have had a 404 Not
  // Found response.
  //
  // This can be useful when the location of a service worker has changed, as
  // it will prevent clients from getting stuck with an old service worker
  // indefinitely.
  //
  // This problem arises because when a service worker updates, a 404 is
  // treated as a failed update. It does not cause the service worker to be
  // unregistered. See https://github.com/w3c/ServiceWorker/issues/204 for more
  // discussion of this problem.
  //
  // Defaults to `true`.
  unregisterMissingServiceWorkers?: boolean;

  // Below is the subset of the polymer.json specification that we care about
  // for serving. https://www.polymer-project.org/2.0/docs/tools/polymer-json
  // https://github.com/Polymer/polymer-project-config/blob/master/src/index.ts
  entrypoint?: string;
  builds?: {
    name?: string,
    browserCapabilities?: capabilities.BrowserCapability[],
  }[];

  // Monitoring configuration for prometheus middleware
  // Basic authentication is optional
  // Default scrape endpoint is `/metrics`
  monitoring?: {
    scrapeEndpoint?: string;
    basicAuth?: {
      username: string,
      password: string
    },
  };
}

// Matches URLs like "/foo/bar.png" but not "/foo.png/bar".
const hasFileExtension = /\.[^/]*$/;

// TODO Service worker location should be configurable.
const isServiceWorker = /service-worker.js$/;

/**
 * Return a new HTTP handler to serve a PRPL-style application.
 */
export function makeHandler(root?: string, config?: Config): (
    request: http.IncomingMessage,
    response: http.ServerResponse,
    next?: express.NextFunction) => void {
  const absRoot = path.resolve(root || '.');
  console.info(`Serving files from "${absRoot}".`);

  const builds = loadBuilds(absRoot, config);

  const cacheControl = (config && config.cacheControl) || 'max-age=60';
  const unregisterMissingServiceWorkers =
      (config && config.unregisterMissingServiceWorkers != undefined) ?
          config.unregisterMissingServiceWorkers :
          true;
  const forwardErrors = config && config.forwardErrors;

  return async function prplHandler(request, response, next) {
    const handleError = (err: httpErrors.HttpError) => {
      if (forwardErrors && next) {
        next(err);
      } else {
        writePlainTextError(response, err);
      }
    };

    response.setHeader('Content-Security-Policy', "default-src * 'unsafe-inline' 'unsafe-eval'; "
        + "script-src * 'unsafe-inline' 'unsafe-eval'; "
        + "connect-src * 'unsafe-inline'; "
        + "font-src * data:; "
        + "img-src * data: blob: 'unsafe-inline'; "
        + "frame-src sanalmarket: yenism: http://*.youtube.com https://tr.rdrtr.com https://stags.bluekai.com https://*.creativecdn.com https://creativecdn.com https://*.criteo.com https://*.facebook.com https://*.doubleclick.net https://*.api.sociaplus.com https://*.webinstats.com https://sanalmarket.api.useinsider.com https://optimize.google.com https://*.bkmexpress.com.tr https://www.linkadoo.co https://linkadoo.co https://channelconnector.smartmessage-connect.com; "
        + "style-src * 'unsafe-inline';");
    response.setHeader('X-Frame-Options', 'SAMEORIGIN');
    response.setHeader('Strict-Transport-Security', 'max-age=0; includeSubDomains');
    response.setHeader('X-XSS-Protection', 1);
    response.setHeader('X-Content-Type-Options', 'nosniff');


    const urlPath = url.parse(request.url || '/').pathname || '/';

    // Let's be extra careful about directory traversal attacks, even though
    // the `send` library should already ensure we don't serve any file outside
    // our root. This should also prevent the file existence check we do next
    // from leaking any file existence information (whether you got the
    // entrypoint or a 403 from `send` might tell you if a file outside our
    // root exists). Add the trailing path separator because otherwise "/foo"
    // is a prefix of "/foo-secrets".
    const absFilepath = path.normalize(path.join(absRoot, urlPath));
    if (!absFilepath.startsWith(addTrailingPathSep(absRoot))) {
      handleError(httpErrors(403, 'Forbidden'));
      return;
    }

    // Serve the entrypoint for the root path, and for all other paths that
    // don't have a corresponding static resource on disk. As a special
    // case, paths with file extensions are always excluded because they are
    // likely to be not-found static resources rather than application
    // routes.
    const serveEntrypoint = urlPath === '/' ||
        (!hasFileExtension.test(urlPath) && !(await fileExists(absFilepath)));

    // Find the highest ranked build suitable for this user agent.
    const clientCapabilities = capabilities.browserCapabilities(
        request.headers['user-agent'] as string);
    const build = builds.find((b) => b.canServe(clientCapabilities));

    // We warned about this at startup. You should probably provide a fallback
    // build with no capabilities, at least to nicely inform the user. Note
    // that we only return this error for the entrypoint; we always serve fully
    // qualified static resources.
    if (!build && serveEntrypoint) {
      handleError(httpErrors(500, 'This browser is not supported.'));
      return;
    }

    const fileToSend = (build && serveEntrypoint) ? build.entrypoint : urlPath;

    if (isServiceWorker.test(fileToSend)) {
      // A service worker may only register with a scope above its own path if
      // permitted by this header.
      // https://www.w3.org/TR/service-workers-1/#service-worker-allowed
      response.setHeader('Service-Worker-Allowed', '/');

      // Don't cache SW (unless cache header is otherwise set).
      if (!response.getHeader('Cache-Control')) {
        response.setHeader('Cache-Control', 'max-age=0');
      }

      // Automatically unregister service workers that no longer exist to
      // prevent clients getting stuck with old service workers indefinitely.
      if (unregisterMissingServiceWorkers && !(await fileExists(absFilepath))) {
        response.setHeader('Content-Type', 'application/javascript');
        response.writeHead(200);
        response.end(
            `self.addEventListener('install', () => self.skipWaiting());
self.addEventListener('activate', () => self.registration.unregister());`);
        return;
      }
    }

    // Don't set the Cache-Control header if it's already set. This way another
    // middleware can control caching, and we won't touch it.
    if (!response.getHeader('Cache-Control')) {
      response.setHeader(
          'Cache-Control', serveEntrypoint ? 'max-age=0' : cacheControl);
    }

    if (build && build.pushManifest) {
      // Set nopush attribute if the client doesn't support push. This will
      // still set preload headers, but it provides a signal to the server to
      // not use server push.
      const nopush = !clientCapabilities.has('push');
      const linkHeaders = build.pushManifest.linkHeaders(urlPath, nopush);
      if (urlPath !== fileToSend) {
        // Also check the filename against the push manifest. In the case of
        // the entrypoint, these will be different (e.g. "/my/app/route" vs
        // "/es2015/index.html"), and we want to support configuring pushes in
        // terms of both.
        linkHeaders.push(...build.pushManifest.linkHeaders(fileToSend, nopush));
      }
      response.setHeader('Link', linkHeaders);
    }

    const sendOpts = {
      root: absRoot,
      // We handle the caching header ourselves.
      cacheControl: false,
    };
    send(request, fileToSend, sendOpts)
        .on('error',
            (err: httpErrors.HttpError) => {
              // `send` puts a lot of detail in the error message, like the
              // absolute system path of the missing file for a 404. We don't
              // want that to leak out, so let's use a generic message instead.
              err.message = statuses.message[err.status] || String(err.status);
              handleError(err);
            })
        .pipe(response);
  };
}

/**
 * Return a promise for the existence of a file.
 */
function fileExists(filepath: string): Promise<boolean> {
  return new Promise((resolve) => fs.access(filepath, (err) => resolve(!err)));
}

/**
 * Write a plain text HTTP error response.
 */
function writePlainTextError(
    response: http.ServerResponse, error: httpErrors.HttpError) {
  response.statusCode = error.status;
  response.setHeader('Content-Type', 'text/plain');
  response.end(error.message);
}

function addTrailingPathSep(p: string): string {
  return p.endsWith(path.sep) ? p : p + path.sep;
}

class Build {
  public pushManifest?: push.PushManifest;

  constructor(
      private configOrder: number,
      public requirements: Set<capabilities.BrowserCapability>,
      public entrypoint: string,
      buildDir: string,
      serverRoot: string) {
    // TODO Push manifest location should be configurable.
    const pushManifestPath = path.join(buildDir, 'push-manifest.json');
    const relPath = path.relative(serverRoot, pushManifestPath);
    if (fs.existsSync(pushManifestPath)) {
      console.info(`Detected push manifest "${relPath}".`);
      // Note this constructor throws if invalid.
      this.pushManifest = new push.PushManifest(
          JSON.parse(fs.readFileSync(pushManifestPath, 'utf8')) as
              push.PushManifestData,
          path.relative(serverRoot, buildDir));
    }
  }

  /**
   * Order builds with more capabililties first -- a heuristic that assumes
   * builds with more features are better. Ties are broken by the order the
   * build appeared in the original configuration file.
   */
  compare(that: Build): number {
    if (this.requirements.size !== that.requirements.size) {
      return that.requirements.size - this.requirements.size;
    }
    return this.configOrder - that.configOrder;
  }

  /**
   * Return whether all requirements of this build are met by the given client
   * browser capabilities.
   */
  canServe(client: Set<capabilities.BrowserCapability>): boolean {
    for (const r of this.requirements) {
      if (!client.has(r)) {
        return false;
      }
    }
    return true;
  }
}

function loadBuilds(root: string, config: Config | undefined): Build[] {
  const builds: Build[] = [];
  const entrypoint = (config ? config.entrypoint : null) || 'index.html';

  if (!config || !config.builds || !config.builds.length) {
    // No builds were specified. Try to serve an entrypoint from the root
    // directory, with no capability requirements.
    console.warn(`WARNING: No builds configured.`);
    builds.push(new Build(0, new Set(), entrypoint, root, root));

  } else {
    for (let i = 0; i < config.builds.length; i++) {
      const build = config.builds[i];
      if (!build.name) {
        console.warn(`WARNING: Build at offset ${i} has no name; skipping.`);
        continue;
      }
      builds.push(new Build(
          i,
          new Set(build.browserCapabilities),
          path.posix.join(build.name, entrypoint),
          path.join(root, build.name),
          root));
    }
  }

  // Sort builds by preference in case multiple builds could be served to
  // the same client.
  builds.sort((a, b) => a.compare(b));

  // Sanity check.
  for (const build of builds) {
    const requirements = Array.from(build.requirements.values());
    console.info(
        `Registered entrypoint "${build.entrypoint}" with capabilities ` +
        `[${requirements.join(',')}].`);
    // Note `build.entrypoint` is relative to the server root, but that's not
    // neccessarily our cwd.
    // TODO Refactor to make filepath vs URL path and relative vs absolute
    // values clearer.
    if (!fs.existsSync(path.join(root, build.entrypoint))) {
      console.warn(`WARNING: Entrypoint "${build.entrypoint}" does not exist.`);
    }
  }
  if (!builds.find((b) => b.requirements.size === 0)) {
    console.warn(
        'WARNING: All builds have a capability requirement. ' +
        'Some browsers will display an error. Consider a fallback build.');
  }

  return builds;
}
